packages/protocol/contracts-0.8/stability/FeeCurrencyAdapter.sol

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.8.7 <0.8.20;
+
+import "@openzeppelin/contracts8/access/Ownable.sol";
+import "@openzeppelin/contracts8/token/ERC20/IERC20.sol";
+
+import "../../contracts/common/CalledByVm.sol";
+import "../../contracts/common/Initializable.sol";
+import "../../contracts/common/interfaces/ICeloVersionedContract.sol";
+import "../../contracts/common/FixidityLib.sol";
+import "../../contracts/stability/interfaces/ISortedOracles.sol";
+import "./interfaces/IFeeCurrency.sol";
+import "./interfaces/IDecimals.sol";
+
+contract FeeCurrencyAdapter is Initializable, CalledByVm {
+  IFeeCurrency public adaptedToken;
+
+  uint96 public digitDifference;
+
+  uint256 public debited;
+
+  string public name;
+  string public symbol;
+
+  event GasFeesDebited(address indexed debitedFrom, uint256 debitedAmount);
+
+  event GasFeesCredited(
+    address indexed refundRecipient,
+    address indexed tipRecipient,
+    address indexed baseFeeRecipient,
+    uint256 refundAmount,
+    uint256 tipAmount,
+    uint256 baseFeeAmount
+  );
+
+  /**
+   * @notice Sets initialized == true on implementation contracts
+   * @param test Set to true to skip implementation initialization
+   */
+  constructor(bool test) public Initializable(test) {}
+
+  /**
+   * @notice Used in place of the constructor to allow the contract to be upgradable via proxy.
+   * @param _adaptedToken The address of the adapted token.
+   * @param _name The name of the adapted token.
+   * @param _symbol The symbol of the adapted token.
+   * @param _expectedDecimals The expected number of decimals of the adapted token.
+   */
+  function initialize(
+    address _adaptedToken,
+    string memory _name,
+    string memory _symbol,
+    uint8 _expectedDecimals
+  ) public virtual initializer {
+    _setAdaptedToken(_adaptedToken);
+    name = _name;
+    symbol = _symbol;
+    uint8 decimals = IDecimals(_adaptedToken).decimals();
+    require(decimals < _expectedDecimals, "Decimals of adapted token must be < expected decimals.");
+    digitDifference = uint96(10**(_expectedDecimals - decimals));
```

Comment:
storage read is cheaper than multiplying inlne?

Reply:
Im not sure that I fully understand the question. Storage read is always necessary, but by storing already multiplied value we dont have to multiply with each credit/debit